### PR TITLE
[Attention] Cache attention metadata builds across hybrid KV-cache groups

### PR DESCRIPTION
**Pull request by LucasWilkinson**
# Essential Elements of an Effective PR Description Checklist

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

## Purpose

Right now when using the hybrid kv-cache manager we build a new attention metadata from scratch for each kv-cache group even when the kv-cache groups only differ by the block-table. This happens we have a repeated n:1 pattern like Llama4 where we have 3 local chunked attention layers for every one full attention layer. In this example all the local attention layer metadatas are virtually identical and only differ by block table. In this PR we cache metadata across kv-cache groups and do a lightweight `update_block_table` instead of full rebuild.

## Test Plan

lm_eval + benchmarking

## Test Result
```
VLLM_ALLOW_CHUNKED_LOCAL_ATTN_WITH_HYBRID_KV_CACHE=1 VLLM_ATTENTION_BACKEND=FLASHINFER python -m lm_eval --model vllm --model_args pretrained=RedHatAI/Llama-4-Scout-17B-16E-Instruct-FP8-dynamic,tensor_parallel_size=2,gpu_memory_utilization=0.8,trust_remote_code=True,max_model_len=32768,disable_hybrid_kv_cache_manager=False,disable_cascade_attn=True --tasks ruler_qa_squad --limit 100 --batch_size auto --metadata='{"max_seq_lengths":[16384]}' 
...
|    Tasks     |Version|Filter|n-shot|Metric|   |Value |   |Stderr|
|--------------|------:|------|-----:|-----:|---|-----:|---|------|
|ruler_qa_squad|      1|none  |     0| 16384|↑  | 0.625|±  |   N/A|
|              |       |none  |     0|  4096|↑  |-1.000|±  |   N/A|
```

```
VLLM_ALLOW_CHUNKED_LOCAL_ATTN_WITH_HYBRID_KV_CACHE=1 python -m lm_eval --model vllm --model_args pretrained=RedHatAI/Llama-4-Scout-17B-16E-Instruct-FP8-dynamic,tensor_parallel_size=2,gpu_memory_utilization=0.8,trust_remote_code=True,max_model_len=32768,disable_hybrid_kv_cache_manager=False,disable_cascade_attn=True --tasks ruler_qa_squad --limit 100 --batch_size auto --metadata='{"max_seq_lengths":[16384]}' 
...
|    Tasks     |Version|Filter|n-shot|Metric|   | Value |   |Stderr|
|--------------|------:|------|-----:|-----:|---|------:|---|------|
|ruler_qa_squad|      1|none  |     0| 16384|↑  | 0.6325|±  |   N/A|
|              |       |none  |     0|  4096|↑  |-1.0000|±  |   N/A|
```

```
branch   variant                  rate   req/s    median TTFT (ms) median TPOT (ms)
------   -------                  ----   -----    ---------------- ----------------       
MAIN     hybrid-kv-cache-enabled  1.0    0.99     74.78            15.31
PR       hybrid-kv-cache-enabled  1.0    0.99     73.62            14.89           
MAIN     hybrid-kv-cache-enabled  10.0   9.84     109.57           38.55                    
PR       hybrid-kv-cache-enabled  10.0   9.84     110.6            37.71           

branch   variant                  rate   req/s    median TTFT (ms) median TPOT (ms)
------   -------                  ----   -----    ---------------- ----------------
MAIN     hybrid-kv-cache-disabled 1.0    0.99     73.14            14.58
PR       hybrid-kv-cache-disabled 1.0    0.99     74.78            14.48         
MAIN     hybrid-kv-cache-disabled 10.0   9.84     109.26           37.25
PR       hybrid-kv-cache-disabled 10.0   9.84     109.89           37.37

Server commands:
MAIN/hybrid-kv-cache-enabled: env VLLM_ALLOW_CHUNKED_LOCAL_ATTN_WITH_HYBRID_KV_CACHE=1 .venv/bin/vllm serve RedHatAI/Llama-4-Scout-17B-16E-Instruct-FP8-dynamic --host localhost --port 3333 --max-model-len 4096 -tp 2 --no-enable-prefix-caching --disable-log-stats --trust-remote-code 
MAIN/hybrid-kv-cache-disabled:  .venv/bin/vllm serve RedHatAI/Llama-4-Scout-17B-16E-Instruct-FP8-dynamic --host localhost --port 3333 --disable-hybrid-kv-cache-manager --max-model-len 4096 -tp 2 --no-enable-prefix-caching --disable-log-stats --trust-remote-code 
PR/hybrid-kv-cache-enabled: env VLLM_ALLOW_CHUNKED_LOCAL_ATTN_WITH_HYBRID_KV_CACHE=1 .venv/bin/vllm serve RedHatAI/Llama-4-Scout-17B-16E-Instruct-FP8-dynamic --host localhost --port 3333 --max-model-len 4096 -tp 2 --no-enable-prefix-caching --disable-log-stats --trust-remote-code 
PR/hybrid-kv-cache-disabled:  .venv/bin/vllm serve RedHatAI/Llama-4-Scout-17B-16E-Instruct-FP8-dynamic --host localhost --port 3333 --disable-hybrid-kv-cache-manager --max-model-len 4096 -tp 2 --no-enable-prefix-caching --disable-log-stats --trust-remote-code 

Client command (template):
CLIENT: vllm bench serve --model RedHatAI/Llama-4-Scout-17B-16E-Instruct-FP8-dynamic --host localhost --port 3333 --dataset-name random --random-input-len 1000 --random-output-len 100 --random-range-ratio 0 --num-prompts <req_rate * 120> --request-rate <req_rate> --save-result --result-dir <results_dir> --result-filename <filename> --ignore-eos --trust-remote-code
```

### CODE REVIEW

File: vllm/attention/layers/chunked_local_attention.py [hunk 1]
```diff
@@ -24,17 +24,33 @@ def create_chunked_local_attention_backend(
...
+        cm, make_block_table = make_local_attention_virtual_batches(
+            attention_chunk_size, cm, block_size)
+        # Save on the metadata_builder for build_postprocess_fn
+        self.make_block_table = make_block_table
+        return cm
+
+    def update_block_table_preprocess_fn(self, metadata,
+                                         blk_table: torch.Tensor,
+                                         slot_mapping: torch.Tensor):
+        return metadata, metadata.make_block_table(blk_table), slot_mapping
```
Comment (heheda12345) [inline vllm/attention/layers/chunked_local_attention.py]:
```suggestion
    x = {}
    def build_preprocess_fn(self, cm: CommonAttentionMetadata):
        cm, make_block_table = make_local_attention_virtual_batches(
            attention_chunk_size, cm, block_size)
        # Save on the metadata_builder for build_postprocess_fn
        x.make_block_table = make_block_table
        return cm

    def update_block_table_preprocess_fn(self, metadata,
                                         blk_table: torch.Tensor,
                                         slot_mapping: torch.Tensor):
        return metadata, x.make_block_table(blk_table), slot_mapping
```

With this `x`(need a better name though), can we remove the `build_postprocess_fn`?

**Comment by LucasWilkinson** [comment]
cc @kushanam @pavanimajety can you please review the FlashInfer pieces

**Comment by heheda12345** [comment]
Can you also try the benchmark script used in vllm performance tracking to see whether the performance degradation is fixed?

File: vllm/v1/attention/backends/flashinfer.py [hunk 2]
```diff
@@ -569,14 +578,62 @@ def build(self,
...
         )
 
         self._plan(attn_metadata)
 
         return attn_metadata
 
+    def update_block_table(self, metadata: FlashInferMetadata,
+                           blk_table: torch.Tensor,
+                           slot_mapping: torch.Tensor):
+        new = copy.copy(metadata)
```
Comment (heheda12345) [inline vllm/v1/attention/backends/flashinfer.py:593]: why are you using deepcopy in flashattn but copy in flashinfer?

File: vllm/attention/layers/chunked_local_attention.py [hunk 3]
```diff
@@ -24,19 +23,30 @@ def create_chunked_local_attention_backend(
     block_size: int,
 ) -> type[AttentionBackend]:
     prefix = f"ChunkedLocalAttention_{attention_chunk_size}_{block_size}_"
-
     underlying_builder = underlying_attn_backend.get_builder_cls()
 
     class ChunkedLocalAttentionBuilder(underlying_builder):  # type: ignore
+        # Make sure the unique_cls_id is not the same for different
+        # attention_chunk_size and block_size.
+        __specialization_key__ = (attention_chunk_size, block_size)
```
Comment (heheda12345) [inline vllm/attention/layers/chunked_local_attention.py]: why do you need it? KVCacheSpec should contain these information.

File: vllm/v1/attention/backends/flashinfer.py (hunk 2 above)
Comment (LucasWilkinson) [reply vllm/v1/attention/backends/flashinfer.py:593]: Good catch; flash attention should a be a copy!

**Comment by LucasWilkinson** [comment]
> Can you also try the benchmark script used in vllm performance tracking to see whether the performance degradation is fixed?

Running here: https://github.com/pytorch/pytorch-integration-testing/actions/runs/17228399056

manually set `VLLM_ALLOW_CHUNKED_LOCAL_ATTN_WITH_HYBRID_KV_CACHE=1` in the PR so we should set that back before landing

File: vllm/attention/layers/chunked_local_attention.py (hunk 3 above)
Comment (LucasWilkinson) [reply vllm/attention/layers/chunked_local_attention.py]: true; actually each dynamically instance is unique so we can just hash on `type(builder)` :+1:

File: vllm/v1/attention/backends/flashinfer.py [hunk 4]
```diff
@@ -508,10 +513,14 @@ def build(self,
                 < block_table_bounds.unsqueeze(1))
         # write self.paged_kv_indices inplace
         num_actual_pages = torch.sum(mask)
+
+        def _update_paged_kv_indices_(paged_kv_indices, block_table_tensor):
+            torch.masked_select(block_table_tensor[:, :max_num_blocks],
+                                mask,
+                                out=paged_kv_indices)
```
Comment (pavanimajety) [inline vllm/v1/attention/backends/flashinfer.py:520]: Curious what the advantage of doing it here in a function is?

File: vllm/v1/attention/backends/flashinfer.py [hunk 5]
```diff
@@ -569,14 +578,62 @@ def build(self,
...
 
         return attn_metadata
 
+    def update_block_table(self, metadata: FlashInferMetadata,
+                           blk_table: torch.Tensor,
+                           slot_mapping: torch.Tensor):
+        new = copy.copy(metadata)
+        # We only made a shallow copy of the metadata, so we need to allocate
+        # a new paged_kv_indices tensor.
+        new.paged_kv_indices = torch.empty_like(metadata.paged_kv_indices)
```
Comment (pavanimajety) [inline vllm/v1/attention/backends/flashinfer.py:596]: Allocations can move to class constructor based on maximum paged_kv_indices and here we can do a tensor slice to the required shape
